apps/api/src/common/authority-checker.service.ts

```diff
@@ -117,28 +126,57 @@ export class AuthorityCheckerService {
       throw new NotFoundException(`Project with id ${entity?.id} not found`)
     }
 
+    const projectAccessLevel = project.accessLevel
     // Get the authorities of the user in the workspace with the project
-    const permittedAuthorities = await getCollectiveProjectAuthorities(
-      userId,
-      project,
-      prisma
-    )
+    const permittedAuthorities: Set<Authority> =
+      await getCollectiveProjectAuthorities(userId, project, prisma)
 
-    // If the user does not have the required authority, or is not a workspace admin, throw an error
-    if (
-      !permittedAuthorities.has(authority) &&
-      !permittedAuthorities.has(Authority.WORKSPACE_ADMIN)
-    ) {
-      throw new UnauthorizedException(
-        `User with id ${userId} does not have the authority in the project with id ${entity?.id}`
-      )
+    if (permittedAuthorities.has(Authority.WORKSPACE_ADMIN)) {
+      return project
+    }
+
+    let canAccessPrivateProject = false
+
+    switch (projectAccessLevel) {
+      case ProjectAccessLevel.GLOBAL:
+        //everyone can access this
+        break
+      case ProjectAccessLevel.INTERNAL:
+        // Any member in the workspace with READ_PROJECT collective authority or the required authority will be able to access the project
+        if (
+          !permittedAuthorities.has(Authority.READ_PROJECT) &&
+          !permittedAuthorities.has(authority)
+        ) {
+          // If the user does not have the required authority or the READ_PROJECT authority and is neither a workspace admin, throw an error
+          throw new UnauthorizedException(
+            `User with id ${userId} does not have the authority in the project with id ${entity?.id}`
+          )
+        }
+        break
+
+      case ProjectAccessLevel.PRIVATE:
+        // Check if the user's role association includes the project's ID in its projectIds field
+        canAccessPrivateProject = await checkUserRoleAssociations(
+          userId,
+          project,
+          authority,
+          prisma
+        )
+
+        // If either the user does not have access or does not have required authority, throw an error
+        if (!canAccessPrivateProject || !permittedAuthorities.has(authority)) {
```

Comment:
In here, we will want to check against `permittedAuthoritiesForProject`. We want to check against the authority passed into the function or `WORKSPACE_ADMIN`.

Reply:
Added checks